### PR TITLE
Test for Python3.6 and minor improvements

### PR DESCRIPTION
**Pull request by funkyfuture**
while testing the 3.6 support i came across some low-hanging fruits to improve readability.

### CODE REVIEW

**Comment by funkyfuture** [comment]
now, that is odd. Python 3.4 shows the same behavior like 3.6. ( i hadn't tested the former locally.)

3.3 is failing due to the absence of the `pathlib` library. do you think keeping to support it is [worth it](https://docs.python.org/devguide/#status-of-python-branches)? i'm neither under the impression that it is much in use. but i can undo that change.

File: .flake8 [hunk 1]
```diff
@@ -1,2 +0,0 @@
-[flake8]
```
Comment (un-def) [inline .flake8:1]: I prefer ‘one tool — one config’ rule.

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
 language: python
+cache: pip
 python:
   - 3.3
   - 3.4
   - 3.5
+  - 3.6
 install:
-  - pip install -r requirements_dev.txt
-  - pip install coveralls
```
Comment (un-def) [inline .travis.yml:8]: You missed to install nose, coverage, and coveralls.

File: polypie.py [hunk 3]
```diff
@@ -13,12 +14,10 @@
 
 
 class PolypieException(Exception):
-
```
Comment (un-def) [inline polypie.py:16]: I prefer one blank line below `class` keyword regardless of the presence a docstring.

File: polypie.py [hunk 4]
```diff
@@ -13,12 +14,10 @@
...
     pass
 
 
 class HashableParameter(Parameter):
-
```
Comment (un-def) [inline polypie.py:21]: I prefer one blank line below `class` keyword regardless of the presence a docstring.

File: polypie.py [hunk 5]
```diff
@@ -32,9 +31,7 @@ def from_parameter(parameter):
         )
 
 
-def _call_func(func_key, args=None, kwargs=None):
-    args = args or ()
-    kwargs = kwargs or {}
+def _call_func(func_key, args=(), kwargs={}):
```
Comment (un-def) [inline polypie.py:34]: Never use mutable object (e.g., `dict` instance) as a default parameter value.

File: polypie.py [hunk 6]
```diff
@@ -55,16 +52,14 @@ def _call_func(func_key, args=None, kwargs=None):
 
 
 def polymorphic(func):
-    global _registry
```
Comment (un-def) [inline polypie.py:58]: Explicit global objects is more obvious.

File: polypie.py [hunk 7]
```diff
@@ -55,16 +52,14 @@ def _call_func(func_key, args=None, kwargs=None):
...
     func_key = func.__module__ + '.' + func.__qualname__
     parameters = signature(func).parameters
     parameters_tuple = tuple(
         HashableParameter.from_parameter(p) for p in parameters.values())
     if func_key not in _registry:
+        @wraps(func)
```
Comment (un-def) [inline polypie.py]: It also changes `__module__` attribute (to `polypie`) and assigns wrapped function to `__wrapped__` attribute.

File: setup.py [hunk 8]
```diff
@@ -1,16 +1,20 @@
...
 
 from setuptools import setup
 
 
-os.chdir(os.path.normpath(os.path.join(os.path.abspath(__file__), os.pardir)))
+PROJECT_DIR = Path(__file__).parent.resolve()
 
 
-with open('README.rst') as f:
+if version_info < (3, 4):
```
Comment (un-def) [inline setup.py:11]: I haven't yet decided to drop Python 3.3 support.

File: .travis.yml [hunk 9]
```diff
@@ -1,12 +1,18 @@
...
+  - travis_retry pip install tox-travis
 script:
-  - nosetests --with-coverage
+  - tox
 after_success:
   coveralls
+
+matrix:
+  allow_failures:
+  - python: 3.6  # FIXME
```
Comment (un-def) [inline .travis.yml:18]: Python 3.3 and 3.4 also fail tests.

**Comment by un-def** [comment]
I'll take a closer look at your proposed changes after fixing main `typing`/`typecheck`-related issue.

File: .travis.yml (hunk 2 above)
Comment (funkyfuture) [reply .travis.yml:8]: nope, tox does that. one tool - one config. ;-)

**Comment by un-def** [comment]
> Python 3.4 shows the same behavior like 3.6.

It happens because `typecheck-decorator` requires (in `setup.py`) the latest (and incompatible) version of `typing` module from PyPI. The same version (I guess) is included in Python 3.6.

File: polypie.py (hunk 5 above)
Comment (funkyfuture) [reply polypie.py:34]: i heard of that, but i don't get it. each time this function gets called, a new object will be created. so, what is worrying about it?

File: polypie.py (hunk 6 above)
Comment (funkyfuture) [reply polypie.py:58]: exactly my thought why i shouldn't remove that when i did it.

**Comment by funkyfuture** [comment]
a'ight, thanks for the review. since much is a matter of gusto or plain wrong, just pick what seems useful to you and dismiss this pr then.

concerning the `__module__` attribute, i'd recommend a test covering that before someone else gets along and tries that; it seemed obviously wrong (in the sense that it could be more elegant) the first time i looked at it.

as a side-note, when this is considered stable, i'd say it belongs to the stdlib's `functools` module.

File: .travis.yml (hunk 2 above)
Comment (un-def) [reply .travis.yml:8]: Yes, you're right, I fogot about `deps`.

**Pull request closed without merging**